### ISSUE TITLE
Ctype.Cannot_apply when destructively substituting type

### ISSUE THREAD

**Issue by smuenzel-js** [ocaml/ocaml#9317]
The following code raises Ctype.Cannot_apply:
```
module type X = sig
  type 'a alpha = 'b constraint 'a = < alpha : 'b>

  type 'a t =
    { alpha : 'a alpha
    }

  module Make( ) : sig
    type 'a pre := 'a t
    type alpha
    type t = < alpha : alpha >
    val global : t pre
  end
end
```
It works fine when using `type 'a pre = 'a t` instead of `type 'a pre := 'a t`

**Comment by smuenzel-js** [comment]
I believe the unification trace that describes this error is (on `val global`):
> Type < alpha : 'a > is not compatible with type t

**Comment by gasche** [comment]
This cannot reproduced with 4.11 and later versions, instead we get:

```
      type 'a pre := 'a t
      ^^^^^^^^^^^^^^^^
Error: Destructive substitutions are not supported for constrained 
       types (other than when replacing a type constructor with 
       a type constructor with the same arguments).
```

I believe that this was fixed by #9388.

Thanks for the report!
